allow showing any content-type in calendar

erral opened the pull request:
I am still testing this (I can't make this addon work in Volto 17 because of some Webpack issues which I don't understand), but I think that with these changes we can make this plugin more flexible and allow showing any content object in the calendar view of a listing provided the integrator/developer implements a function with the information to show on it.

erral committed:
allow showing any content-type in calendar

erral committed:
ups

erral commented:
I have been testing this and it works perfectly in Volto 17 after fixing the webpack issues (See #12)

erral committed:
fix config import

erral committed:
add docs

erral committed:
improve readme

stevepiercy commented on README.md:
````diff
@@ -23,6 +23,29 @@ https://user-images.githubusercontent.com/4497578/154287664-6153ae22-27dd-48ed-9
 ```
 More details checkout the following link: https://5.docs.plone.org/external/plone.app.dexterity/docs/advanced/catalog-indexing-strategies.html#adding-new-indexes-and-metadata-columns
 
+### Add content-types other than Events in the Calendar
````
```suggestion
### Add content types other than events in the calendar
```

stevepiercy commented on README.md:
```diff
@@ -23,6 +23,29 @@ https://user-images.githubusercontent.com/4497578/154287664-6153ae22-27dd-48ed-9
 
+### Add content-types other than Events in the Calendar
+
+You may have developed a custom version of an Event that want to show in a calendar, or you may have other content-types that want to show in a calendar.
```
```suggestion
You may have developed a custom version of an event or other content types that want to show in a calendar.
```

stevepiercy commented on README.md:
```diff
@@ -23,6 +23,29 @@ https://user-images.githubusercontent.com/4497578/154287664-6153ae22-27dd-48ed-9
+
+You may have developed a custom version of an Event that want to show in a calendar, or you may have other content-types that want to show in a calendar.
+
+To do so, you need to provide a function that returns the date, title and urls to show for your content-type.
```
```suggestion
To do so, you need to provide a function that returns the date, title, and URLs to show for your content type.
```

stevepiercy commented on README.md:
```diff
@@ -23,6 +23,29 @@ https://user-images.githubusercontent.com/4497578/154287664-6153ae22-27dd-48ed-9
+
+To do so, you need to provide a function that returns the date, title and urls to show for your content-type.
+
+You need to register that function in the block config in your addon this way:
```
```suggestion
You need to register that function in the block configuration in your add-on this way:
```

stevepiercy commented on README.md:
````diff
@@ -23,6 +23,29 @@ https://user-images.githubusercontent.com/4497578/154287664-6153ae22-27dd-48ed-9
+
+```
+
+The block already includes a default function for the Event content-type.
````
```suggestion
The block already includes a default function for the event content type.
```

stevepiercy reviewed (changes requested):
Minor English grammar and syntax fixes.

erral committed:
Update README.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

erral committed:
Update README.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

erral committed:
Update README.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

erral committed:
Update README.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

erral committed:
Update README.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

erral commented:
Thanks @stevepiercy!

stevepiercy reviewed (approved):
Docs look good. Someone else should take a look at the code.

ksuess commented on src/components/manage/Blocks/Listing/FullCalendar.jsx:
```diff
@@ -59,9 +60,13 @@ const FullCalendarListing = ({ items, moment: momentlib, ...props }) => {
   let events = items
     .filter((i) => {
-      if (i['@type'] !== 'Event') return false;
+      if (!Object.keys(contentConverters).includes(i['@type'])) return false;
+
```
Why not filtering by event like types, which are items with 'start' and 'end', then skipping the introduction of a 'contentConverters' function?

erral replied:
The `contentConverters` provides a flexible way to let the developer decide which kind of items he wants to show in the calendar.

For instance, I have items that are not events, that don't have 'start' and 'end' attributes. That's why I built this function, as if it was a ZCA adapter, this way you can decide which fields you want to use as `title, `start` and `end` and this way show the item on the calendar.

ksuess replied:
IMHO it's the job of the backend to provide common practice attributes like 'start' and 'end' according IEventBasics behavior.
By now you are asking the integrator to provide a configuration for any other content type than 'event'. There should be a fallback for IEventBasic behavior enhanced content types. They do have a 'start' attribute.

erral replied:
Well, right now the existing implementation only supports Event content-types (like in the deleted condition). I have kept the implementation for Events, and extend it for any other content-type.

ksuess replied:
Your solution lacks a default / fallback for instances that already provide 'start', for example  IEventBasics enhanced content types.

Again: I would prefer to only loosen the filter condition  from `if (i['@type'] !== 'Event')` to just `if (!i.start)`. No extra configuration. But instead on backend providing custom practice attributes of the IEventBasics.

erral replied:
ah Ok. I wasn't understanding your request. That's doable of course.

erral committed:
extract filtering and converstion to dedicated functions

erral commented:
I have extracted the filtering and the conversion to dedicated functions and implemented the default behavior for items having `start` and `end` attributes as suggested by @ksuess

stevepiercy commented:
@erral it looks the docs are still OK with this change. Can you verify?

ksuess commented:
This is not a collective repo, so it's up to the repo owner @mbarde to accept a catch fix for a poor backend implementation.

IMHO Event like objects do have IEventBasic attributes. 
An opening to event like objects is OK and needed. 
This just needs a change in the filter condition  from 
`if (i['@type'] !== 'Event')`
to 
`if (!i.start)`

erral commented:
I don't think that my implementation is poor because I am not using IEventBasic attributes in my content-type.

That implementation is **my** choice and I am **not** forced to use IEventBasic behavior to model my content-type.

At the same time, the maintainer **is not forced to accept** this PR.

I think I have provided a very extensible implementation to allow showing anything on the calendar view, not being tied to use IEventBasic (which I didn't think of initially). I have also updated the PR with your suggestion, allowing items that provide `start` and `end`.

Ok. I am not going to argue more on this. If the repo owner is happy, fair enough. If not, I don't mind. I have my branch where I can keep working on.

mbarde commented:
I agree with @ksuess and removed the type restriction in the newest release (1.0.0) as suggested (https://github.com/mbarde/volto-fullcalendar-block/commit/ea7d012a0b9e593bcfd564a786936340a6e6aede).

I wont add "official" support for content-types which do not implement IEventBasic. 

@erral A suggested approach to go without a custom fork: Shadow the  `Listing/FullCalendar`  component and build a wrapper around the original component which transforms `items` and feeds the transformed array into the original (see https://community.plone.org/t/volto-composition-to-customize-components/14059/3).

(I will have to do this for a project within the next days, where we have the requirement that single objects should result in multiple events in the calendar. So if you are interested I could provide a code example for this.)

Furthermore,  I added support for easily passing custom props to the FullCalendar component, which could useful to cover special use-cases as well: https://github.com/mbarde/volto-fullcalendar-block?tab=readme-ov-file#customize-fullcalendar-options

mbarde commented:
@erral Possible way to do data transformation without modifying the original components: https://github.com/mbarde/volto-fullcalendar-block?tab=readme-ov-file#customize-listing-block-variation
